### PR TITLE
Restructure instruction fusion code

### PR DESCRIPTION
**Pull request by amiri-khalil**

### CODE REVIEW

File: lbr.py [hunk 1]
```diff
@@ -338,23 +342,25 @@ def is_imix(t):
 Insts_global = Insts + is_imix(None) + ['all']
 Insts_cond = ['backward-taken', 'forward-taken', 'non-taken', 'fusible', 'non-fusible', 'taken-not-first'
               ] + ['%s-JCC non-fusible'%x for x in user_jcc_pair]
-Insts_all = ['cond_%s'%x for x in Insts_cond] + Insts_global
+Insts_MF2 = [x + '-OP fusible' for x in ['MOV', 'LD']]
```
Comment (aayasin) [inline lbr.py:345]: rename Insts_MF2 to Insts_Fusions e.g.

File: lbr.py [hunk 2]
```diff
@@ -403,6 +409,38 @@ def header_only_str(l):
     if loop_ipc:
       read_sample.tick *= 10
       read_sample.stop = None
+  def edge_stats():
```
Comment (aayasin) [inline lbr.py:412]: Thanks for adding this function.
Please hoist it outside read_sample() and pass what's needed, like lines, to it.

File: lbr.py [hunk 3]
```diff
@@ -733,19 +744,22 @@ def print_hist(hist_t, Threshold=0.01):
...
     if x in is_imix(None):  n += 'insts-class'
     elif 'cond' in name:    n += 'branches'
+    elif 'fusible' in name: n += 'pairs'
     else: n += 'instructions'
     return n
   if len(comment): comment = '\t:(see %s below)' % c(comment)
   elif ratio_of: comment = '\t: %7s of %s' % (ratio(count, ratio_of[1]), ratio_of[0])
-  print('%s of %s: %10s%s' % (c(prefix), '{: >{}}'.format(c(nm(name)), 60 - len(prefix)), str(count), comment))
+  res = '%s of %s: %10s%s' % (c(prefix), '{: >{}}'.format(c(nm(name)), 60 - len(prefix)), str(count), comment)
+  if log: C.fappend(res, log)
```
Comment (aayasin) [inline lbr.py:761]: how about 1 line:

C.fappend(res, log) if log else print(res)

File: stats.py [hunk 4]
```diff
@@ -297,6 +299,74 @@ def basename():
   print('wrote:', stat)
   return stat
 
+def print_candidate_stats(hitcounts, info):
```
Comment (aayasin) [inline stats.py:302]: I would rename it to inst_fusions()
I wouldn't stress 'print' and stats is module's name.

File: stats.py [hunk 5]
```diff
@@ -297,6 +299,74 @@ def basename():
...
+        else:  # basic block end, check candidates
+          for i, block_line in enumerate(block):
+            if condition(block_line): find_cand(block[i:])
+          if condition(line):
+            hotness_key, block = hotness(line), [line]  # restart for next block
+          else: hotness_key, block = None, None
+  total = int(C.exe_one_line(C.grep(' ALL instructions:', info)).split(':')[1].strip())
+  calc_stats()
+  for stat, value in stats_data.items():
+    print_stat('%s fusible-candidate' % stat, value, ratio_of=('ALL', total), log=info)
```
Comment (aayasin) [inline stats.py:368]: Great code reuse :)

File: stats.py [hunk 6]
```diff
@@ -297,6 +299,74 @@ def basename():
   print('wrote:', stat)
   return stat
 
+def print_candidate_stats(hitcounts, info):
+  stats_data = {'LD-OP': 0,
+                'MOV-OP': 0}
+  def calc_stats():
+    block = hotness_key = None
+    hotness = lambda s: C.str2list(s)[0]
+    condition = lambda l: 'mov' in l and not x86.is_mem_store(l)
```
Comment (aayasin) [inline stats.py:308]: renaming condition to is_MOV make helps readability.

File: stats.py [hunk 7]
```diff
@@ -297,6 +299,74 @@ def basename():
...
+          block = [line]
+          hotness_key = hotness(line)
+          continue
+        # append lines from the same basic block (by hotness)
+        if hotness(line) == hotness_key: block.append(line)
+        else:  # basic block end, check candidates
+          for i, block_line in enumerate(block):
+            if condition(block_line): find_cand(block[i:])
+          if condition(line):
+            hotness_key, block = hotness(line), [line]  # restart for next block
```
Comment (aayasin) [inline stats.py:363]: how about making this a tiny restart() local function that is invoked here, and in lines 354-355, and also for next line when overriding default args, for example.

**Review by aayasin — CHANGES_REQUESTED** [review]
Amiri, enclosed more review comments to be addressed.

File: kernels/x86.py [hunk 8]
```diff
@@ -30,11 +30,16 @@
...
-STORE     = r"\s+\S+\s+[^\(\),]+,"
-MOVE      = r"v?mov"
+STORE     = r"\s+\S+\s+[^\(\),]+,"  # use is_mem_store()
+MOV       = r"v?mov"
+CMOV      = r"cmov"
 BR = '(j|%s|sys%s|bnd jmp)' % (CALL_RET, CALL_RET)
 MEM_IDX = r"\((%[a-z0-9]+)?,%[a-z0-9]+,?(1|2|4|8)?\)"
 M_FUSION_INSTS = ['cmp', 'test', 'add', 'sub', 'inc', 'dec', 'and']
+REGS_32 = ['eax', 'ebx', 'ecx', 'edx', 'esi', 'edi', 'ebp', 'esp'] + \
+          ['r' + str(n) + 'd' for n in range(8, 16)]
```
[the remark this answers is not in the document]
Comment (amiri-khalil) [reply kernels/x86.py:40]: Yes, notice the 'd' there, e.g. r8d is the 32-bit reg for r8, and so on.

**Review by aayasin — APPROVED** [review]
Thanks for addressing online review comments

**Pull request merged**